pkg/virt-api/webhooks/validating-webhook/admitters/vmi-update-admitter.go

```diff
@@ -64,6 +78,115 @@ func (admitter *VMIUpdateAdmitter) Admit(ar *v1beta1.AdmissionReview) *v1beta1.A
+			},
+		})
+	}
+	hotplugVolumeMap := getHotplugVolumes(newVolumes, volumeStatuses)
+	permanentVolumeMap := getPermanentVolumes(newVolumes, hotplugVolumeMap)
+
+	// Ensure we didn't remove non hot plugged disks and volumes.
+	permanentCount := 0
+	for _, volume := range newVolumes {
+		if _, ok := permanentVolumeMap[volume.Name]; ok {
+			permanentCount++
+		}
+	}
+	if len(permanentVolumeMap) > permanentCount || len(permanentVolumeMap) == 0 {
+		// Removed one of the permanent volumes, reject admission.
+		return webhookutils.ToAdmissionResponse([]metav1.StatusCause{
+			{
+				Type:    metav1.CauseTypeFieldValueInvalid,
+				Message: "cannot remove permanent volume",
+			},
+		})
+	}
+	// Ensure the permanent volumes and disks are the first in the list.
+	for i := 0; i < len(permanentVolumeMap); i++ {
+		if _, ok := permanentVolumeMap[newVolumes[i].Name]; !ok {
+			return webhookutils.ToAdmissionResponse([]metav1.StatusCause{
+				{
+					Type:    metav1.CauseTypeFieldValueInvalid,
+					Message: fmt.Sprintf("Volume %s is not after the permanent volumes", newVolumes[i].Name),
+				},
+			})
+		}
+		if _, ok := permanentVolumeMap[newDisks[i].Name]; !ok {
+			return webhookutils.ToAdmissionResponse([]metav1.StatusCause{
+				{
+					Type:    metav1.CauseTypeFieldValueInvalid,
+					Message: fmt.Sprintf("Disk %s is not after the permanent disks", newDisks[i].Name),
+				},
+			})
+		}
+	}
+
+	// Ensure all the disks and volumes have matching names.
+	for _, disk := range newDisks {
+		if _, ok := hotplugVolumeMap[disk.Name]; !ok {
+			// Not a hotplug volume, check permanent volumes
+			if _, ok := permanentVolumeMap[disk.Name]; !ok {
+				// Not in persistent map either, this disk doesn't have a matching volume.
+				return webhookutils.ToAdmissionResponse([]metav1.StatusCause{
+					{
+						Type:    metav1.CauseTypeFieldValueInvalid,
+						Message: fmt.Sprintf("Disk %s doesn't have a matching volume", disk.Name),
+					},
+				})
+			}
+		} else {
+			// Ensure the volume source is either PVC or DataVolume
+			volume, _ := hotplugVolumeMap[disk.Name]
+			if volume.DataVolume == nil && volume.PersistentVolumeClaim == nil {
+				return webhookutils.ToAdmissionResponse([]metav1.StatusCause{
+					{
+						Type:    metav1.CauseTypeFieldValueInvalid,
+						Message: fmt.Sprintf("Disk %s has a volume that is not a PVC or DataVolume", disk.Name),
+					},
+				})
+			}
+		}
+	}
+	return nil
+}
+
+func getHotplugVolumes(volumes []v1.Volume, volumeStatuses []v1.VolumeStatus) map[string]v1.Volume {
+	hotplugVolumes := make(map[string]v1.Volume, 0)
+	for _, volume := range volumeStatuses {
+		if volume.Hotplug != nil {
+			hotplugVolumes[volume.Name] = v1.Volume{}
+		}
+	}
+	for _, volume := range volumes {
+		if _, ok := hotplugVolumes[volume.Name]; ok {
+			hotplugVolumes[volume.Name] = volume
+		}
+	}
+	// Make sure the map only contains valid volumes
+	for k, v := range hotplugVolumes {
+		if k != v.Name {
```

Comment:
the only way this condition would be met is if ```v.Name == ""```, right?

Reply:
Correct

Reply:
Its for the case we deleted a volume, which is still in the statuses (we haven't updated it yet)